File: platform/src/lib/agents/llm-client.ts
import Anthropic from "@anthropic-ai/sdk";
import { GoogleGenerativeAI } from "@google/generative-ai";

export const GEMINI_MODEL = "gemini-3.7-flash";
export const HAIKU_MODEL = "claude-haiku-4-5-20251001";

export interface GenerateArgs {
  system: string;
  prompt: string;
  maxTokens: number;
}

export interface GenerateResult {
  text: string;
  model: string;
  promptTokens?: number;
  completionTokens?: number;
}

const CALL_TIMEOUT_MS = 90_000;

function withTimeout<T>(promise: Promise<T>, label: string, ms = CALL_TIMEOUT_MS): Promise<T> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error(`${label} timed out after ${ms}ms`)),
      ms
    );
    promise.then(
      (value) => {
        clearTimeout(timer);
        resolve(value);
      },
      (err) => {
        clearTimeout(timer);
        reject(err);
      }
    );
  });
}

/**
 * Primary/fallback model routing for the agent pipeline: Gemini 3.7 Flash
 * first (fast, cheap — the roster runs 9+ sequential calls per engagement),
 * falling back to Claude Haiku 4.5 if Gemini is unavailable, errors, or no
 * GEMINI_API_KEY is configured. Callers never need to know which model
 * actually answered — the returned `model` field records it for the audit
 * trail (persisted per phase run).
 */
export async function generateText({
  system,
  prompt,
  maxTokens,
}: GenerateArgs): Promise<GenerateResult> {
  const geminiKey = process.env.GEMINI_API_KEY;

  if (geminiKey) {
    try {
      return await withTimeout(
        generateWithGemini(geminiKey, { system, prompt, maxTokens }),
        "Gemini call"
      );
    } catch (err) {
      console.error(
        `[llm-client] Gemini call failed, falling back to Haiku: ${
          err instanceof Error ? err.message : String(err)
        }`
      );
    }
  }

  const anthropicKey = process.env.ANTHROPIC_API_KEY;
  if (!anthropicKey) {
    throw new Error(
      "Neither GEMINI_API_KEY nor ANTHROPIC_API_KEY is set. Add at least one to platform/.env.local."
    );
  }

  return withTimeout(
    generateWithHaiku(anthropicKey, { system, prompt, maxTokens }),
    "Haiku call"
  );
}

async function generateWithGemini(
  apiKey: string,
  { system, prompt, maxTokens }: GenerateArgs
): Promise<GenerateResult> {
  const genAI = new GoogleGenerativeAI(apiKey);
  const model = genAI.getGenerativeModel({
    model: GEMINI_MODEL,
    systemInstruction: system,
  });

  const result = await model.generateContent({
    contents: [{ role: "user", parts: [{ text: prompt }] }],
    generationConfig: { maxOutputTokens: maxTokens },
  });

  const text = result.response.text();
  if (!text.trim()) {
    throw new Error("Gemini returned an empty response.");
  }

  const usage = result.response.usageMetadata;
  return {
    text,
    model: GEMINI_MODEL,
    promptTokens: usage?.promptTokenCount,
    completionTokens: usage?.candidatesTokenCount,
  };
}

async function generateWithHaiku(
  apiKey: string,
  { system, prompt, maxTokens }: GenerateArgs
): Promise<GenerateResult> {
  const anthropic = new Anthropic({ apiKey });

  const message = await anthropic.messages.create({
    model: HAIKU_MODEL,
    max_tokens: maxTokens,
    system,
    messages: [{ role: "user", content: prompt }],
  });

  const textBlock = message.content.find((b) => b.type === "text");
  const text = textBlock && textBlock.type === "text" ? textBlock.text : "";

  return {
    text,
    model: HAIKU_MODEL,
    promptTokens: message.usage.input_tokens,
    completionTokens: message.usage.output_tokens,
  };
}

File: engine/fde_engine/llm_client.py
"""Every model call the pipeline makes goes through here.

Ported from platform/src/lib/agents/llm-client.ts. Gemini 3.7 Flash first —
fast and cheap, and the roster runs ten sequential calls per engagement —
falling back to Claude Haiku 4.5 when Gemini errors or no key is configured.
Callers never need to know which model answered; the returned `model` records
it for the audit trail, persisted per phase run.
"""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass

log = logging.getLogger(__name__)

GEMINI_MODEL = "gemini-3.7-flash"
HAIKU_MODEL = "claude-haiku-4-5-20251001"

#: Without a timeout a hung request has no way to fail, and a stalled pipeline
#: looks identical to a slow one.
CALL_TIMEOUT_SECONDS = 90.0


@dataclass(frozen=True)
class GenerateResult:
    text: str
    model: str
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    #: True when the model hit the output ceiling and the text is cut off
    #: mid-thought.
    truncated: bool = False


class NoModelKeyError(RuntimeError):
    """Neither provider is configured."""


def generate_text(*, system: str, prompt: str, max_tokens: int) -> GenerateResult:
    """Primary/fallback routing for the agent pipeline."""
    gemini_key = os.environ.get("GEMINI_API_KEY")
    if gemini_key:
        try:
            return _with_gemini(
                gemini_key, system=system, prompt=prompt, max_tokens=max_tokens
            )
        except Exception as err:  # noqa: BLE001 — any Gemini failure falls back
            log.warning("Gemini call failed, falling back to Haiku: %s", err)

    anthropic_key = os.environ.get("ANTHROPIC_API_KEY")
    if not anthropic_key:
        raise NoModelKeyError(
            "Neither GEMINI_API_KEY nor ANTHROPIC_API_KEY is set. "
            "Add at least one to platform/.env.local."
        )

    return _with_haiku(
        anthropic_key, system=system, prompt=prompt, max_tokens=max_tokens
    )


def _with_gemini(
    api_key: str, *, system: str, prompt: str, max_tokens: int
) -> GenerateResult:
    from google import genai
    from google.genai import types

    client = genai.Client(api_key=api_key, http_options={"timeout": int(CALL_TIMEOUT_SECONDS * 1000)})
    response = client.models.generate_content(
        model=GEMINI_MODEL,
        contents=prompt,
        config=types.GenerateContentConfig(
            system_instruction=system,
            max_output_tokens=max_tokens,
        ),
    )

    text = response.text or ""
    if not text.strip():
        # Gemini 3.x spends part of max_output_tokens on internal reasoning
        # ("thoughts"). A tight budget can be consumed entirely by thinking,
        # leaving no visible answer at all.
        raise RuntimeError(
            f"Gemini returned an empty response (max_tokens={max_tokens} "
            "may have been consumed by reasoning tokens)."
        )

    usage = response.usage_metadata
    candidates = response.candidates or []
    finish = getattr(candidates[0], "finish_reason", None) if candidates else None

    return GenerateResult(
        text=text,
        model=GEMINI_MODEL,
        prompt_tokens=getattr(usage, "prompt_token_count", None),
        completion_tokens=getattr(usage, "candidates_token_count", None),
        truncated=str(finish).endswith("MAX_TOKENS"),
    )


def _with_haiku(
    api_key: str, *, system: str, prompt: str, max_tokens: int
) -> GenerateResult:
    import anthropic

    client = anthropic.Anthropic(api_key=api_key, timeout=CALL_TIMEOUT_SECONDS)
    message = client.messages.create(
        model=HAIKU_MODEL,
        max_tokens=max_tokens,
        system=system,
        messages=[{"role": "user", "content": prompt}],
    )

    text = next((b.text for b in message.content if b.type == "text"), "")

    return GenerateResult(
        text=text,
        model=HAIKU_MODEL,
        prompt_tokens=message.usage.input_tokens,
        completion_tokens=message.usage.output_tokens,
        truncated=message.stop_reason == "max_tokens",
    )
